utils/s2n_fork_detection.c

```diff
@@ -0,0 +1,369 @@
+     *
+     * Instead, we apply best-effort to initialise the MADV_WIPEONFORK fork
+     * detection and otherwise always require pthread_atfork to be initialised.
+     * We also currently always apply prediction resistance. So, this should be
+     * a safe default.
+     */
+    if (ignore_wipeonfork_or_inherit_zero_method_for_testing == S2N_FORK_DETECT_DO_NOT_IGNORE) {
+        s2n_initialise_wipeonfork_best_effort(addr, page_size);
+    }
+
+    if (ignore_wipeonfork_or_inherit_zero_method_for_testing == S2N_FORK_DETECT_DO_NOT_IGNORE) {
+        POSIX_GUARD(s2n_initialise_inherit_zero(addr, page_size));
+    }
+
+    if (ignore_pthread_atfork_method_for_testing == S2N_FORK_DETECT_DO_NOT_IGNORE) {
+        POSIX_GUARD(s2n_inititalise_pthread_atfork());
+    }
+
+    fgn_state.zero_on_fork_addr = addr;
+    *fgn_state.zero_on_fork_addr = S2N_NO_FORK_EVENT;
+    fgn_state.is_fork_detection_enabled = S2N_FORK_DETECT_ENABLED;
+
+    return S2N_SUCCESS;
+}
+
+static void s2n_initialise_fork_detection_methods(void)
+{
+    void *addr = MAP_FAILED;
+    long page_size = 0;
+
+    /* Only used to disable fork detection mechanisms during testing. */
+    if (ignore_wipeonfork_or_inherit_zero_method_for_testing == S2N_FORK_DETECT_IGNORE &&
+        ignore_pthread_atfork_method_for_testing == S2N_FORK_DETECT_IGNORE) {
+
+        ignore_fork_detection_for_testing = S2N_FORK_DETECT_IGNORE;
+        return;
+    }
+
+    page_size = sysconf(_SC_PAGESIZE);
+    if (page_size <= 0) {
+        return;
+    }
+
+    addr = mmap(NULL, (size_t) page_size, PROT_READ | PROT_WRITE,
+                MAP_PRIVATE | MAP_ANONYMOUS, -1, 0);
+    if (addr == MAP_FAILED) {
+        return;
+    }
+
+    /* Now we know that we have some memory mapped. Try to initialise fork
+     * detection methods. Unmap the memory if we fail for some reason!
+     */
+    if (s2n_initialise_fork_detection_methods_try(addr, page_size) != S2N_SUCCESS) {
+        /* No reason to verify return value of munmap() since we can't use that
+         * information for anything anyway. */
+        munmap(addr, (size_t) page_size);
+        addr = NULL;
+        fgn_state.zero_on_fork_addr = NULL;
+        fgn_state.is_fork_detection_enabled = S2N_FORK_DETECT_NOT_ENABLED;
+    }
+}
+
+/* s2n_get_fork_generation_number returns S2N_RESULT_OK on success and
+ * S2N_RESULT_ERROR otherwise.
+ *
+ * On success, returns the current fork generation number in
+ * return_fork_generation_number. Caller must synchronise access to
+ * return_fork_generation_number.
+ */
+S2N_RESULT s2n_get_fork_generation_number(uint64_t *return_fork_generation_number)
+{
+    RESULT_ENSURE(pthread_once(&fgn_state.fork_detection_once, s2n_initialise_fork_detection_methods) == 0, S2N_ERR_FORK_DETECTION_INIT);
+
+    if (ignore_fork_detection_for_testing == S2N_FORK_DETECT_IGNORE) {
+        /* Fork detection is meant to be disabled. Hence, return success. */
+        return S2N_RESULT_OK;
+    }
```

Comment:
Because it would be bad if this somehow got set outside of testing, do you think it's worth repeating the "we are in a test" assertion here?